rust/signer/src/lib.rs

```diff
@@ -236,15 +296,83 @@ export! {
     }
 }
 
+secure_native::export_put! {
+    @Java_io_parity_signer_EthkeyBridge_snPut
+    fn sn_put(success: Result<(), String>) -> Result<(), String> {
+        success
+    }
+}
+
+secure_native::export_get! {
+    @Java_io_parity_signer_EthkeyBridge_snEthkeyBrainwalletSign
+    fn sn_ethkey_brainwallet_sign(pin: Result<String, String>, message: String, encrypted: String) -> Result<String, String> {
+        let seed = decrypt(&encrypted, &pin?)?;
+        let phrase = parse_suri(&seed)?.phrase;
+        let (_, keypair) = KeyPair::from_auto_phrase(&phrase);
+        let message: Vec<u8> = message.from_hex().map_err(|e| format!("{}", e))?;
+        let signature = keypair.sign(&message).map_err(|e| format!("{}", e))?;
+        Ok(signature.to_hex())
+    }
+
+    @Java_io_parity_signer_EthkeyBridge_snSubstrateBrainwalletSign
+    fn sn_substrate_brainwallet_sign(pin: Result<String, String>, path: String, password: String, message: String, encrypted: String, legacy: bool) -> Result<String, String> {
+        let seed = decrypt(&encrypted, &pin?)?;
+        let suri = if legacy { seed } else {
+            let phrase = parse_suri(&seed)?.phrase;
+            construct_suri(&phrase, &path, &password)
+        };
+        let keypair = sr25519::KeyPair::from_suri(&suri).ok_or(format!("KeyPair from suri"))?;
+        let message: Vec<u8> = message.from_hex().map_err(|e| format!("{}", e))?;
+        let signature = keypair.sign(&message);
+        Ok(signature.to_hex())
```

Comment:
Maybe a dumb question, at this point, does rust auto garbage collect `keypair`?

Reply:
Yes, it gets destroyed when it leaves scope (when the function returns)